azure-functions-codegen/src/export.rs

```diff
@@ -1,45 +1,50 @@
+use crate::util::MacroError;
 use crate::util::{PathVec, TryFrom};
 use proc_macro::TokenStream;
 use quote::quote;
-use std::fmt::Write;
-use syn::{parse_str, Expr};
+use syn::{parse_str, spanned::Spanned, Path};
 
 pub fn attr_impl(input: TokenStream) -> TokenStream {
-    let funcs = match PathVec::try_from(input.clone()) {
-        Ok(funcs) => funcs,
+    let (mods, funcs) = match PathVec::try_from(input.clone()) {
+        Ok(paths) => {
+            let mut mods: Vec<_> = Vec::new();
+            let mut funcs = Vec::new();
+            for path in paths.into_iter() {
+                if path.leading_colon.is_some() || path.segments.len() > 1 {
+                    let error: MacroError = (
+                        path.span(),
+                        "fully qualified names are not supported for the `export` macro",
+                    )
+                        .into();
+
+                    error.emit();
+                    return input;
+                }
+
+                let segment = path.segments.first().unwrap();
+                let identifier = segment.value().ident.to_string();
+
+                funcs.push(
+                    parse_str::<Path>(&format!(
+                        "{}::__{}_FUNCTION",
+                        identifier,
+                        identifier.to_uppercase()
+                    ))
+                    .unwrap(),
+                );
+                mods.push(path);
+            }
+            (mods, funcs)
+        }
         Err(e) => {
             e.emit();
             return input;
         }
     };
-    let funcs: Vec<Expr> = funcs
-        .into_iter()
-        .map(|path| {
-            let mut expr = String::new();
-            if path.leading_colon.is_some() {
-                expr += "::";
-            }
-
-            let mut segments = path.segments.into_iter().peekable();
-            while let Some(segment) = segments.next() {
-                if segments.peek().is_some() {
-                    write!(&mut expr, "{}::", segment.ident).unwrap();
-                } else {
-                    write!(
-                        &mut expr,
-                        "__{}_FUNCTION",
-                        segment.ident.to_string().to_uppercase()
-                    )
-                    .unwrap();
-                }
-            }
-
-            parse_str::<Expr>(&expr).unwrap()
-        })
-        .collect();
 
     let expanded = quote! {
-        &[#(&#funcs),*]
+        #(mod #mods;)*
+        pub const FUNCTIONS: &[&::azure_functions::codegen::Function] = &[#(&#funcs),*];
```

Comment:
It would be nice if `codegen` wasn't even available on `azure_functions`. I still like the idea of changing `azure_functions` to re-export `Function`. I know this isn't directly visible by the user, but it could cause confusion later on.

Reply:
There's a lot more exposed than just `Function`.  `#[func]` generates a lot of code that uses `codegen`.

Yeah, it'd be nice if users couldn't see the "behind the curtain" module, but I'd prefer not to pollute the root with a bunch of types that users will never use.

Reply:
That's fair. I think it's something to keep an eye on.